Mathlib/Util/Qq.lean

```diff
@@ -35,4 +35,22 @@ def findLocalDeclWithTypeQ? {u : Level} (sort : Q(Sort u)) : MetaM (Option Q($so
   let some fvarId ← findLocalDeclWithType? q($sort) | return none
   return some <| .fvar fvarId
 
+/-- Returns the natural number from a natural literal expression, or `none` if the expression isn't
+a natural literal.
+
+See `Lean.Expr.rawNatLit?` for a `MetaM`-less version of this function that only checks for raw
+literals. -/
+def natLitQq? : Q(Nat) → MetaM (Option Nat)
+  | ~q(OfNat.ofNat $n) => pure n.rawNatLit?
+  | e => pure e.rawNatLit?
```

Comment:
While this seems like a nice enough thing to have, we already have a [baffling number of these](https://loogle.lean-lang.org/?q=Lean.Expr%20%E2%86%92%20Option%20Nat), and I don't see why your use case needs your new one when previous ones didn't.

Reply:
Uh that's terrible. They all look misnamed

Reply:
Do you agree that it would be worth unifying all these?

Reply:
I think they all do various amounts of post-processing (raw / non-raw, numeral / succ / zero, etc).

For the purpose of the first version your simproc, I think it's fine to support only positive numerals for now, where you can just copy the approach I used for the prime factors simproc.